model/src/main/kotlin/Provenance.kt

```diff
@@ -80,7 +80,9 @@ data class RepositoryProvenance(
     val resolvedRevision: String
 ) : KnownProvenance {
     init {
-        require(resolvedRevision.isNotBlank()) { "The resolved revision must not be blank." }
+        require(vcsInfo == VcsInfo.EMPTY || resolvedRevision.isNotBlank()) {
```

Comment:
Commit message nits:
- s/vcs/VCS/
- s/git/Git/
- Odd indentation before "Both"
- Again use passive voice

Reply:
Logic-wise, I'd prefer to have the clause order swapped. Usually we require `resolvedRevision.isNotBlank()` but if it is blank, the whole `VcsInfo` must be empty: `resolvedRevision.isNotBlank() || vcsInfo == VcsInfo.EMPTY`.

Reply:
>  * s/vcs/VCS/

`vcs` is part of a test case name string. so changing the capitalization would erode the reference.
Maybe I should change the format to `"fail if no vcs matches"` to make this more clear and also explicitly call them test cases?

Reply:
Swapped the logic around.